File: School/CP164/hust4880_data_structures/src/Food_utilities.py
"""
-------------------------------------------------------
Food class utility functions.
-------------------------------------------------------
Author:  David Brown
ID:      999999999
Email:   dbrown@wlu.ca
Section: CP164 Winter 2020
__updated__ = "2020-01-10"
-------------------------------------------------------
"""
from Food import Food
from copy import deepcopy


def get_food():
    """
    -------------------------------------------------------
    Creates a food object by requesting data from a user.
    Use: f = get_food()
    -------------------------------------------------------
    Returns:
        food - a completed food object (Food).
    -------------------------------------------------------
    """
    Name = input("Name: ")
    print(Food.origins())
    Origin = int(input(": "))
    Veg = ''
    
    Veg = input('Vegetarian (Y/N): ').lower()
    if(Veg == 'y'):
        Veg = True
    else:
        Veg = False
        
    Calories = input("Calories: ")
    if(Calories == 'None'):
        Calories = None
    else:
        Calories = int(Calories)

    food = Food(Name, Origin, Veg, Calories)
    return food


def read_food(line):
    """
    -------------------------------------------------------
    Creates and returns a food object from a line of string data.
    Use: f = read_food(line)
    -------------------------------------------------------
    Parameters:
        line - a vertical bar-delimited line of food data in the format
          name|origin|is_vegetarian|calories (str)
    Returns:
        food - contains the data from line (Food)
    -------------------------------------------------------
    """

    food_v = line.split('|')
    if(food_v[2] == 'True'):
        food_v[2] = True
    elif(food_v[2] == 'False'):
        food_v[2] = False
        
    for i in food_v:
        print(i)
        
    food = Food(food_v[0],int(food_v[1]),food_v[2],int(food_v[3]))

    return food


def read_foods(file_variable):
    """
    -------------------------------------------------------
    Reads a file of food strings into a list of Food objects.
    Use: foods = read_food(file_variable)
    -------------------------------------------------------
    Parameters:
        file_variable - a file of food data (file)
    Returns:
        foods - a list of food objects (list of Food)
    -------------------------------------------------------
    """
    counter = 0
    foods = []
    for line in file_variable.readlines():
        
        counter += 1
        
        food = line.strip('\n').split('|')

        if(food[2] == 'True'):
            food[2] = True
        elif(food[2] == 'False'):
            food[2] = False
            
        foods.append(Food(food[0], int(food[1]), food[2], int(food[3])))
    
    return foods


def write_foods(file_variable, foods):
    """
    -------------------------------------------------------
    Writes a list of food objects to a file.
    file_variable contains the objects in foods as strings in the format
          name|origin|is_vegetarian|calories
    Use: write_foods(file_variable, foods)
    -------------------------------------------------------
    Parameters:
        file_variable - an open file of food data (file)
        foods - a list of Food objects (list of Food)
    Returns:
        None
    -------------------------------------------------------
    """
    for i in foods:
        string = '{}|{}|{}|{}\n'.format(i.name,i.origin,i.is_vegetarian,i.calories)
        file_variable.write(string)

    return


def get_vegetarian(foods):
    """
    -------------------------------------------------------
    Creates a list of vegetarian foods.
    Use: v = get_vegetarian(foods)
    -------------------------------------------------------
    Parameters:
        foods - a list of Food objects (list of Food)
    Returns:
        veggies - Food objects from foods that are vegetarian (list of Food)
    -------------------------------------------------------
    """

    veggies = []
    for i in foods:
        if(i.is_vegetarian):
            veggies.append(i)
            
    return veggies


def by_origin(foods, origin):
    """
    -------------------------------------------------------
    Creates a list of foods by origin.
    Use: v = by_origin(foods, origin)
    -------------------------------------------------------
    Parameters:
        foods - a list of Food objects (list of Food)
        origin - a food origin (int)
    Returns:
        origins - Food objects from foods that are of a particular origin (list of Food)
    -------------------------------------------------------
    """
    assert origin in range(len(Food.ORIGIN))
    origins = []
    
    for i in foods:
        if(i.origin == origin):
            origins.append(i)
    

    return origins


def average_calories(foods):
    """
    -------------------------------------------------------
    Determines the average calories in a list of foods.
    Use: avg = average_calories(foods)
    -------------------------------------------------------
    Parameters:
        foods - a list of Food objects (list of Food)
    Returns:
        avg - average calories in all Food objects of foods (int)
    -------------------------------------------------------
    """
    total = 0
    for i in foods:
        total += i.calories
    avg = total//len(foods)
    return avg


def calories_by_origin(foods, origin):
    """
    -------------------------------------------------------
    Determines the average calories in a list of foods.
    Use: a = calories_by_origin(foods, origin)
    -------------------------------------------------------
    Parameters:
        foods - a list of Food objects (list of Food)
        origin - the origin of the Food objects to find (int)
    Returns:
        avg - average calories for all Foods of the requested origin (int)
    -------------------------------------------------------
    """
    assert origin in range(len(Food.ORIGIN))
    total = 0
    count = 0
    for i in foods:
        if(i.origin == origin):
            total += i.calories
            count += 1
    avg = total//count

    return avg


def food_table(foods):
    """
    -------------------------------------------------------
    Prints a formatted table of foods, sorted by name.
    Use: food_table(foods)
    -------------------------------------------------------
    Parameters:
        foods - a list of Food objects (list of Food)
    Returns:
        None
    -------------------------------------------------------
    """
    

    print("""
Food                                Origin       Vegetarian Calories
----------------------------------- ------------ ---------- --------""")
    foods_copy = deepcopy(foods)
    foods_copy.sort()

    for i in foods_copy:
        print('{:<36}{:<13}{:<10}{:<11}'.format(i.name, Food.ORIGIN[i.origin],str(i.is_vegetarian),i.calories))

    return 



def food_search(foods, origin, max_cals, is_veg):
    """
    -------------------------------------------------------
    Searches for foods that fit certain conditions.
    Use: results = food_search(foods, origin, max_cals, is_veg)
    -------------------------------------------------------
    Parameters:
        foods - a list of Food objects (list of Food)
        origin - the origin of the food; if -1, accept any origin (int)
        max_cals - the maximum calories for the food; if 0, accept any calories value (int)
        is_veg - whether the food is vegetarian or not; if False accept any food (boolean)
    Returns:
        result - a list of foods that fit the conditions (list of Food)
            foods parameter must be unchanged
    -------------------------------------------------------
    """
    assert origin in range(-1, len(Food.ORIGIN))
    
    result = []
    for i in foods:
        if((i.origin == origin or origin == -1) and (i.calories <= max_cals or max_cals == -1) and (i.is_vegetarian or not is_veg)):
            result.append(i)
    return result

File: School/CP213/hust4880_a02/src/cp213/FoodUtilities.java
package cp213;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.Scanner;

/**
 * Utilities for working with Food objects.
 *
 * @author your name here
 * @version 2021-01-29
 */
public class FoodUtilities {

    /**
	 * Determines the average calories in a list of foods. No rounding necessary.
	 * Foods list parameter may be empty.
	 *
	 * @param foods a list of Food
	 * @return average calories in all Food objects
	 */
	public static int averageCalories(final ArrayList<Food> foods) {
		int total = 0;
		int avg = 0;
		for(Food f : foods)
			total += f.getCalories();
		
		avg = total/foods.size();
		return avg;
	}

    /**
     * Determines the average calories in a list of foods for a particular origin.
     * No rounding necessary. Foods list parameter may be empty.
     *
     * @param foods  a list of Food
     * @param origin the origin of the Food
     * @return average calories for all Foods of the specified origin
     */
    public static int averageCaloriesByOrigin(final ArrayList<Food> foods, final int origin) {
    	int total = 0;
    	int food_number = 0;
    	int avg = 0;
    	for(Food f : foods) {
    		if(f.getOrigin() == origin) {
    			total += f.getCalories();
    			food_number++;
    		}
    	}
    	avg = total/food_number;
    	return avg;
    }

    /**
     * Creates a list of foods by origin.
     *
     * @param foods  a list of Food
     * @param origin a food origin
     * @return a list of Food from origin
     */
    public static ArrayList<Food> getByOrigin(final ArrayList<Food> foods, final int origin) {
    	ArrayList<Food> ByOrigin = new ArrayList<Food>();
    	for(Food f : foods) {
    		if(f.getOrigin() == origin)
    			ByOrigin.add(f);
    	}
    	return ByOrigin;
    }

    /**
     * Creates a Food object by requesting data from a user. Uses the format:
     * 
     * <pre>
    Name: name
    Origins
     0 Canadian
     1 Chinese
    ...
    11 English
    Origin: origin number
    Vegetarian (Y/N): Y/N
    Calories: calories
     * </pre>
     *
     * @param keyboard a keyboard Scanner
     * @return a Food object
     */
    public static Food getFood(final Scanner keyboard) {
    	boolean isVeg = false;
    	String name;
    	int origin;
    	int calories;
    	System.out.print("Name: ");
    	name = keyboard.nextLine();
    	
    	System.out.print(Food.originsMenu());
    	System.out.print("Origin: ");
    	origin = keyboard.nextInt();
    	
    	System.out.print("Vegetarian (Y/N): ");
    	String yn = keyboard.next();
    	if(yn.toLowerCase().equals("y"))
    		isVeg = true;
    	else if(yn.toLowerCase().equals("n"))
    		isVeg = false;
    	
    	System.out.print("Calories: ");
    	calories = keyboard.nextInt();
    	
    	
    	Food new_food = new Food(name, origin, isVeg, calories);
    	return new_food;
    }

    /**
     * Creates a list of vegetarian foods.
     *
     * @param foods a list of Food
     * @return a list of vegetarian Food
     */
    public static ArrayList<Food> getVegetarian(final ArrayList<Food> foods) {
    	ArrayList<Food> veglist = new ArrayList<Food>();
    	for(Food f : foods) {
    		if(f.isVegetarian() == true)
    			veglist.add(f);
    	}
    	return veglist;
    }

    /**
     * Creates and returns a Food object from a line of string data.
     *
     * @param line a vertical bar-delimited line of food data in the format
     *             name|origin|isVegetarian|calories
     * @return the data from line as a Food object
     */
    public static Food readFood(final String line) {
    	String[] food_data = line.split("\\|");
    	
    	String name = food_data[0];
    	int origin = Integer.parseInt(food_data[1]);
    	boolean isVeg;
    	if(food_data[2].toLowerCase().equals("true"))
    		isVeg = true;
    	else
    		isVeg = false;
    	int calories = Integer.parseInt(food_data[3]);
    	
    	Food new_food = new Food(name, origin, isVeg, calories);
    	return new_food;
    }

    /**
     * Reads a file of food strings into a list of Food objects.
     *
     * @param fileIn a Scanner of a Food data file in the format
     *               name|origin|isVegetarian|calories
     * @return a list of Food
     */
    public static ArrayList<Food> readFoods(final Scanner fileIn) {
    	ArrayList<Food> foods = new ArrayList<Food>();
    	String line = "";
    	while(fileIn.hasNextLine()) {
        	line = fileIn.nextLine();
        	foods.add(FoodUtilities.readFood(line));
    	}
    	return foods;
    }

    /**
     * Searches for foods that fit certain conditions.
     *
     * @param foods        a list of Food
     * @param origin       the origin of the food; if -1, accept any origin
     * @param maxCalories  the maximum calories for the food; if 0, accept any
     * @param isVegetarian whether the food is vegetarian or not; if false accept
     *                     any
     * @return a list of foods that fit the conditions specified
     */
    public static ArrayList<Food> foodSearch(final ArrayList<Food> foods, final int origin, final int maxCalories,
	    final boolean isVegetarian) {
    	
    	ArrayList<Food> Searched = new ArrayList<Food>();
    	if(isVegetarian) {
	    	for(Food f : foods) {
	    		if((f.getOrigin() == origin || origin == -1) && f.getCalories() <= maxCalories && f.isVegetarian())
	    			Searched.add(f);
	    	}
    	}
    	else
	    	for(Food f : foods) {
	    		if((f.getOrigin() == origin || origin == -1) && f.getCalories() <= maxCalories)
	    			Searched.add(f);
	    	}
    	return Searched;
    	
    }

    /**
     * Writes the contents of a list of Food to a PrintStream.
     *
     * @param foods a list of Food
     * @param ps    the PrintStream to write to
     */
    public static void writeFoods(final ArrayList<Food> foods, PrintStream ps) {
    	for(Food f : foods) 
    		f.write(ps);
    }
}